services-core/src/solution_submission/retry.rs

```diff
@@ -24,76 +15,97 @@ use std::{
 
 const GAS_PRICE_REFRESH_INTERVAL: Duration = Duration::from_secs(15);
 
-pub struct Args {
-    pub batch_index: u32,
-    pub solution: Solution,
-    pub claimed_objective_value: U256,
-    pub gas_price_cap: f64,
-    pub nonce: U256,
-    pub target_confirm_time: Instant,
+#[async_trait::async_trait]
+#[cfg_attr(test, mockall::automock(type Output = bool;))]
+pub trait TransactionSending: Send {
+    type Output: TransactionResult;
+    async fn send(&self, gas_price: f64) -> Self::Output;
+}
+
+pub trait TransactionResult {
+    /// Was the transaction mined or did it error before? This happens when a transaction was
+    /// replaced with a higher gas price.
+    fn was_mined(&self) -> bool;
+}
+
+// For mocking we picked the associated type to be bool so it must implement this trait.
+#[cfg(test)]
+impl TransactionResult for bool {
+    fn was_mined(&self) -> bool {
+        *self
+    }
 }
 
-#[derive(Debug)]
-pub enum RetryResult {
-    Submitted(Result<(), MethodError>),
-    Cancelled(Result<(), ExecutionError>),
+pub enum RetryResult<TransactionResult, CancellationResult> {
+    Submitted(TransactionResult),
+    Cancelled(CancellationResult),
 }
 
-#[cfg_attr(test, mockall::automock)]
-pub trait SolutionTransactionSending: Send + Sync {
+#[async_trait::async_trait]
+pub trait RetryTransactionSending: Send + Sync {
     /// Submit the solution with an appropriate gas price based on target_confirm_time. Until the
     /// transaction has been confirmed the gas price is continually updated.
     /// When cancel_after is ready the transaction will be cancelled by sending a noop transaction
     /// at a higher gas price.
-    fn retry<'a>(
-        &'a self,
-        args: Args,
-        cancel_after: BoxFuture<'static, ()>,
-    ) -> BoxFuture<'a, RetryResult>;
+    /// Returns None if no transaction has been submitted.
+    async fn retry<TransactionSender, CancellationSender>(
+        &self,
+        transaction_sender: TransactionSender,
+        cancel_after: BoxFuture<'_, CancellationSender>,
```

Comment:
I made the CancellationSender be returned from the future because it allows users of the function who don't care about cancellation to use a future::pending without having to create a CancellationSender.